### ISSUE TITLE
can not install magenta in google colaboratory

### ISSUE THREAD

**Issue by changhyunkwang** [magenta/magenta#2081]
If i take code !pip install magenta

------------

result:

---------

Collecting magenta==2.1.0
  Using cached magenta-2.1.0-py3-none-any.whl (1.4 MB)
Requirement already satisfied: absl-py in /usr/local/lib/python3.10/dist-packages (from magenta==2.1.0) (1.4.0)
Collecting apache-beam>=2.14.0 (from apache-beam[gcp]>=2.14.0->magenta==2.1.0)
  Using cached apache_beam-2.51.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (6.1 kB)
Collecting dm-sonnet (from magenta==2.1.0)
  Using cached dm_sonnet-2.0.1-py3-none-any.whl (268 kB)
Requirement already satisfied: imageio in /usr/local/lib/python3.10/dist-packages (from magenta==2.1.0) (2.31.5)
Requirement already satisfied: librosa>=0.6.2 in /usr/local/lib/python3.10/dist-packages (from magenta==2.1.0) (0.7.2)
Requirement already satisfied: matplotlib>=1.5.3 in /usr/local/lib/python3.10/dist-packages (from magenta==2.1.0) (3.7.1)
Collecting mido==1.2.6 (from magenta==2.1.0)
  Using cached mido-1.2.6-py2.py3-none-any.whl (69 kB)
Collecting mir-eval>=0.4 (from magenta==2.1.0)
  Using cached mir_eval-0.7.tar.gz (90 kB)
  Preparing metadata (setup.py) ... done
Collecting note-seq (from magenta==2.1.0)
  Using cached note_seq-0.0.5-py3-none-any.whl (209 kB)
Collecting numba<0.50 (from magenta==2.1.0)
  Using cached numba-0.49.1.tar.gz (2.0 MB)
  Preparing metadata (setup.py) ... done
Requirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from magenta==2.1.0) (1.23.0)
Requirement already satisfied: Pillow>=3.4.2 in /usr/local/lib/python3.10/dist-packages (from magenta==2.1.0) (9.4.0)
Collecting pretty-midi>=0.2.6 (from magenta==2.1.0)
  Using cached pretty_midi-0.2.10.tar.gz (5.6 MB)
  Preparing metadata (setup.py) ... done
Collecting pygtrie>=2.3 (from magenta==2.1.0)
  Using cached pygtrie-2.5.0-py3-none-any.whl (25 kB)
Collecting python-rtmidi<1.2,>=1.1 (from magenta==2.1.0)
  Using cached python-rtmidi-1.1.2.tar.gz (204 kB)
  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  Preparing metadata (setup.py) ... error
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.

-------

also i check
https://github.com/magenta/magenta/issues/2072
https://github.com/magenta/magenta/issues/2045
https://github.com/magenta/magenta/issues/2031
this page
but i can't fix the bug.

**Comment by Lynnn96** [comment]
Did you solve this issue? I also run into the same problem as you did and can't fix it.

**Comment by kmylavar25** [comment]
+1 Same here as well. I have the same problem as well. None of the suggestions I found online worked for me.

**Comment by belk4cura** [comment]
Having the same issue

**Comment by changhyunkwang** [comment]
> Did you solve this issue? I also run into the same problem as you did and can't fix it.

No, I think the problem is with the Tensorflow or other modules installed in the Google colaboratory.

**Comment by kmylavar25** [comment]
I installed python 3.7 and that solved the issue for me. I couldn't get the colab notebooks to work, but I was able to successfully install the magenta package. I installed pyenv on my mac and got this working.

**Comment by jmoso13** [comment]
> > Did you solve this issue? I also run into the same problem as you did and can't fix it.
> 
> No, I think the problem is with the Tensorflow or other modules installed in the Google colaboratory.

This is likely the issue, CoLab updated their python version awhile ago and may have done again recently, this causes a lot of old ML packages and dependencies to no longer work, I had the same issue trying to use Jukebox awhile back in CoLab. Really annoying they don't allow for multiple versions of python or virtual environments easily.

Just checked stack overflow and looks like there are some people who have reported luck manually downgrading python version, haven't been able to accomplish it personally though.

**Comment by yeeking** [comment]
I found a really hacky way to run magenta in colab. It also works for Ubuntu and WSL with nearly the same commands. The basis of it is to install python3.7, then create a virtual env inside there. For colab, you then use the xterm extension to pop up a terminal and interact with your python 3.7
 
(https://github.com/InfuseAI/colab-xterm)

Here is a screenshot: 

![image](https://github.com/magenta/magenta/assets/26504/77304052-df99-499e-8942-3ec33cd1bea6)

Here is a colab: 

https://colab.research.google.com/drive/1gDcjWFS2EgXHJS2lFo8NJQV1f8FDjwbU?usp=sharing

These are the commands - you can run these on a regular ubuntu:

```
# https://stackoverflow.com/questions/52816156/how-to-create-virtual-environment-for-python-3-7-0
sudo add-apt-repository ppa:deadsnakes/ppa -y
sudo apt-get install -y  build-essential libasound2-dev libjack-dev portaudio19-dev python3.7 python3.7-venv python3.7-dev
python3.7 -m venv ~/Python/mag37
source ~/Python/mag37/bin/activate
pip install magenta
pip install google.colab # if on colab only I think
# https://stackoverflow.com/questions/41415629/importerror-no-module-named-tensorflow-python
pip uninstall tensorflow
pip install tensorflow

```

**Comment by olaviinha** [comment]
Change `'python-rtmidi == 1.1.2',` in `setup.py` to `'python-rt-midi==1.5.8'` prior to installing. That worked for me. 

Or just [use this repo](https://github.com/olaviinha/magenta). I don't actively maintain it but every now and then I fix Colab issues in there as they come along, so still better maintained than this repo lol.

**Comment by zhanh-he** [comment]
> Clone from repository and change `'python-rtmidi == 1.1.2',` in `setup.py` to `'python-rt-midi==1.5.8'` prior to installing. That worked for me.
> 
> Or just [use this repo](https://github.com/olaviinha/magenta). I don't actively maintain it but every now and then I fix Colab issues in there as they come along, so still more actively maintained than this repo lol.

Good to have this, but not enough. Magenta is requiring TF2.9.1 & Numpy1.21.6 which are not supported by Colab anymore. We need to downgrade Colab to Python3.7 & CUDA11.7 & GCC9 to make it fully run. Python3.7 is for numpy==1.21.6; CUDA & GCC is for the GPU (ignore it if you happy with CPU).

Attached my Colab Notebook for a quick start (https://colab.research.google.com/drive/1AP8SqXH58M-Pcj4KxJbK_A0xGE8OAidM?usp=sharing).

**Comment by saviornt** [comment]
The issue is with the required libraries that are installed with Magenta, specifically numba, python-rtmidi and llvmlite. All of which have required versions of python which are much older (as noted above, they require the use of Python 3.7). Hopefully Google will update the project to use the current versions of said libraries and Python.
